### DISCUSSION TITLE
Boxplot - wrong logic

### DISCUSSION THREAD

**Discussion by li-ana** [apache/superset#18423] [Ideas]
Box plot assumes that you will have only 1 observation per timestamp. In my case, this is not true, which means that the box plot will aggregate all of the observations per timestamp using the function that you select, thus it skews the results. It needs to look at raw data, instead of the aggregation. 

Query produced by the Superset box plot:
**
![image](https://user-images.githubusercontent.com/20177485/101823239-79851f80-3af8-11eb-917b-79121d8004ee.png)
**
Results of this query:
![image](https://user-images.githubusercontent.com/20177485/101823445-c2d56f00-3af8-11eb-8a27-4dafc6a98459.png)

Box plot details:
![image](https://user-images.githubusercontent.com/20177485/101823581-fca67580-3af8-11eb-90ef-8444e4dad961.png)

What it should be:
![image](https://user-images.githubusercontent.com/20177485/101823662-1e076180-3af9-11eb-9883-5e0eb787252d.png)

**Comment by villebro** [comment]
While we don't yet support using the full raw data and then calculating the boxplot on that, we recently migrated Boxplot to ECharts and added some features in this PR: #11199 . In the below example I've created a Boxplot where the categories are continents and the distribution is calculated across countries (I'm using the average of total population, as the dataset contains data for multiple years):

![image](https://user-images.githubusercontent.com/33317356/101879303-cc62e380-3b99-11eb-9793-a82eb52575e8.png)

If you have a row id, you can use that as the the "Distribute Across" parameter. The plan is to add support for using the raw row data, but I probably won't have time to work on it any time soon.

**Comment by rumbin** [comment]
The problem with calculating the Boxplot metrics on the query result is - even if distributing across a unique column - that the row limit hits hard and silently:
If the number of data points across all series exceeds the row limit, the resulting boxplot is non-deterministically excluding data points without notifying the user.
Non-deterministically, since there is no ORDER BY applied, nor is it configurable.

So we have three issues that are caused by the current Boxplot logic:
1. There is no way of including all records, as soon as the number of rows exceeds the row limit.
2. Whether the row limit has been reached is not displayed anywhere.
3. The row limit excludes records in a non-deterministic fashion, as no explicit ordering is present.

In my eyes, all of these issues can best be covered by calculating all Boxplot metrics per series directly within the SQL query. The only drawback that I can immediately see is that outliers cannot be returned by such a query...

**Comment by rumbin** [comment]
Not sure to what extent it would be useful to create new issues for the three items above...

**Comment by rumbin** [comment]
I filed a separate bug for it: #17042

**Comment by junlincc** [comment]
@rumbin please feel free to open separate issues for all.
for 2. it's happening on all the charts, no?

**Comment by rumbin** [comment]
@junlincc 
For 2.: Yes and no.
True is that hitting the row limit is not shown on charts on a dashboard.
However, at least in Explore the row count indicator pill is turning red when the threshold has been reached.
This is _normally_ the case, bit not so for the Box Plot.
That's what I have reported in #17942.

**Comment by rumbin** [comment]
I am going to write up a separate issue for 1. soon, which will elaborate the pros and cons of calculating the box plot metrics in a push down fashion directly in the database...
